src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -12197,75 +12067,42 @@ private InitializerExpressionSyntax ParseObjectOrCollectionInitializer()
         {
             var openBrace = this.EatToken(SyntaxKind.OpenBraceToken);
 
-            var initializers = _pool.AllocateSeparated<ExpressionSyntax>();
-            this.ParseObjectOrCollectionInitializerMembers(ref openBrace, initializers, out var kind);
-            Debug.Assert(initializers.Count > 0 || kind == SyntaxKind.ObjectInitializerExpression);
+            var initializers = this.ParseCommaSeparatedSyntaxList(
+                ref openBrace,
+                SyntaxKind.CloseBraceToken,
+                static @this => @this.IsInitializerMember(),
+                static @this => @this.ParseObjectOrCollectionInitializerMember(),
+                static (@this, openBrace, list, expectedKind, _) => @this.SkipBadInitializerListTokens(openBrace, list, expectedKind),
+                allowTrailingSeparator: true);
+
+            var kind = isObjectInitializer(initializers) ? SyntaxKind.ObjectInitializerExpression : SyntaxKind.CollectionInitializerExpression;
```

Comment:
the old logic would try to maintain state while parsing the elements.  the new logic doesn't, keeping things much cleaner.